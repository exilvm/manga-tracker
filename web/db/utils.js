const format = require('pg-format');
const {
  NUMERIC_VALUE_OUT_OF_RANGE,
  INVALID_TEXT_REPRESENTATION,
} = require('pg-error-constants');
const dblog = require('debug')('db');

function generateEqualsColumns(o, availableColumns) {
  const cols = new Set(availableColumns);
  const sqlValues = [];
  const columns = [];
  const args = [];

  Object.keys(o)
    .filter(k => o[k] !== undefined && cols.has(k))
    .forEach((k, idx) => {
      sqlValues.push(`%I=$${idx+1}`);
      columns.push(k);
      args.push(o[k]);
    });

  const sqlCols = format(sqlValues.join(','), ...columns);

  return { sqlCols, args };
}
module.exports.generateEqualsColumns = generateEqualsColumns;

function handleError(err, res) {
  if (err.code === INVALID_TEXT_REPRESENTATION) {
    dblog(err.message);
    res.status(400).json({ error: 'Invalid data type given' });
    return;
  }

  if (err.code === NUMERIC_VALUE_OUT_OF_RANGE) {
    res.status(400).json({ error: 'Number value out of range' });
    return;
  }

  console.error(err);
  res.status(500).json({ error: 'Internal server error' });
}
module.exports.handleError = handleError;
