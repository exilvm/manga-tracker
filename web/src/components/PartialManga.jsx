import React from 'react';
import {Grid, makeStyles, Tooltip, Typography,} from '@material-ui/core';
import MangaSourceList from "./MangaSourceList";

const dateOptions = { weekday: 'short', year: 'numeric', month: 'long', day: 'numeric', hour: '2-digit', minute: '2-digit' };

const useStyles = makeStyles((theme) => ({
  title: {
    width: '100%',
    textAlign: 'left',
    paddingBottom: '10px',
    [theme.breakpoints.down('sm')]:{
      ...theme.typography.h6,
    }
  },
  thumbnail: {
    maxWidth: '200px',
    maxHeight: '300px',
    [theme.breakpoints.down('sm')]: {
      maxWidth: '125px'
    }
  },
  details: {
    display: 'flex',
  },
  detailText: {
    marginLeft: '5px'
  },
  infoTable : {
    marginLeft: theme.spacing(3),
    marginTop: '3px',
  },
  sourceList: {
    marginLeft: theme.spacing(3),
  },
  paper: {
    padding: '1em',
  }
}));

function PartialManga(props) {
  const {
    showId = false,
    mangaData,
  } = props;

  if (!mangaData.manga_id) return null;

  const classes = useStyles();
  const latest_release = mangaData.latest_release ?
    new Date(mangaData.latest_release) :
    null;

  const estimated_release = mangaData.estimated_release ?
    new Date(mangaData.estimated_release) :
    null;

  return (
    <div>
      <Typography className={classes.title} variant='h5'>{mangaData.title}</Typography>
      <div className={classes.details}>
        <a href={mangaData.mal} target='_blank'>
          <img
            src={mangaData.cover}
            className={classes.thumbnail}
            alt={mangaData.title}
          />
        </a>
        <Grid
            container
            justify='space-between'
            direction='column'
        >
          <table className={classes.infoTable}>
            <tbody>
            { showId &&
              <tr>
                <td><Typography>Manga id:</Typography></td>
                <td><Typography className={classes.detailText}>{mangaData.manga_id}</Typography></td>
              </tr>
            }
              <tr>
                <td><Typography>Latest release:</Typography></td>
                <td>
                  <Tooltip title={latest_release ? latest_release.toUTCString() : 'Unknown'}>
                    <Typography className={classes.detailText}>
                      {latest_release ? latest_release.toLocaleString('en-GB', dateOptions) : 'Unknown'}
                    </Typography>
                  </Tooltip>
                </td>
              </tr>
              <tr>
                <td><Typography>Estimated release interval:</Typography></td>
                <td>
                  <Typography className={classes.detailText}>
                    {(mangaData.release_interval ?
                        `${mangaData.release_interval?.days || 0} days ${mangaData.release_interval?.hours || 0} hours`
                        : 'Unknown')}
                  </Typography>
                </td>
              </tr>
              <tr>
                <td><Typography>Estimated next release:</Typography></td>
                <td>
                  <Typography className={classes.detailText}>
                    {estimated_release ? estimated_release.toLocaleString('en-GB', dateOptions) : 'Unknown'}
                  </Typography>
                </td>
              </tr>
            <tr>
                <td><Typography>Latest chapter:</Typography></td>
                <td>
                  <Typography className={classes.detailText}>
                    {mangaData.latest_chapter ? mangaData.latest_chapter : 'Unknown'}
                  </Typography>
                </td>
              </tr>
            </tbody>
          </table>
          <div className={classes.sourceList}>
            <MangaSourceList
              items={mangaData.services}
              isAuthenticated={false}
            />
          </div>
        </Grid>

      </div>
    </div>
  );
}

export default PartialManga;