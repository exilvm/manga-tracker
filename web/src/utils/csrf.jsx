import { createContext, useContext } from 'react';

const CSRFContext = createContext(undefined);
export const CSRFProvider = CSRFContext.Provider;

export const useCSRF = () => useContext(CSRFContext);

/**
 * Returns the useCSRF token in the props property with the key _csrf
 * @param req
 * @returns {{props: {_csrf: String}}}
 */
export const csrfProps = ({ req }) => ({
  props: { _csrf: req.csrfToken() },
});

export const csrfHeader = (csrf) => ({
  'X-CSRF-Token': csrf,
});
