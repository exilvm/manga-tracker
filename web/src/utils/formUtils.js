export const showErrorAlways = ({
  meta: { submitError, dirtySinceLastSubmit, error },
}) => !!(((submitError && !dirtySinceLastSubmit) || error));
