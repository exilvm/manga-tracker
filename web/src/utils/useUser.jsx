import { createContext, useContext, useMemo } from 'react';


const UserContext = createContext(undefined);
export const UserProvider = UserContext.Provider;

export const useUser = () => {
  const user = useContext(UserContext);
  return useMemo(() => ({
    user,
    isAuthenticated: !!user,
    isAdmin: user && user.admin,
  }), [user]);
};
