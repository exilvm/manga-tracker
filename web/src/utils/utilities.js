import { format, formatDistanceToNowStrict } from 'date-fns';
import enLocale from 'date-fns/locale/en-GB';
import throttle from 'lodash.throttle';


// eslint-disable-next-line import/prefer-default-export
export const followUnfollow = (mangaId, serviceId) => {
  const url = serviceId ? `/api/user/follows?manga_id=${mangaId}&service_id=${serviceId}` :
                           `/api/user/follows?manga_id=${mangaId}`;
  return throttle((event) => {
    const target = event.target;
    switch (target.textContent.toLowerCase()) {
      case 'follow':
        fetch(url, { credentials: 'include', method: 'put' })
            .then(res => {
              if (res.status === 200) {
                target.textContent = 'Unfollow';
              }
            });
        break;

      case 'unfollow':
        fetch(url, { credentials: 'include', method: 'delete' })
            .then(res => {
              if (res.status === 200) {
                target.textContent = 'Follow';
              }
            });
        break;

      default:
        target.textContent = 'Follow';
    }
  }, 200, { trailing: false });
};

// This seems to be faster than a custom recursive function  according to my measurements
export function jsonSerializable(value) {
  return JSON.parse(JSON.stringify(value));
}

export function defaultDateFormat(date, ifUndefined='Unknown') {
  if (!date || Number.isNaN(date.getTime())) return ifUndefined;

  format(date, 'MMM do, HH:mm', { locale: enLocale });
}

export function defaultDateDistanceToNow(date, ifUndefined='Unknown') {
  if (!date || Number.isNaN(date.getTime())) return ifUndefined;

  return formatDistanceToNowStrict(date, { addSuffix: true });
}
