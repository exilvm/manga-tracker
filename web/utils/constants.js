const csrfMissing = 'Missing CSRF token';
const isDev = process.env.NODE_ENV !== 'production';
const isTest = process.env.NODE_ENV === 'test';


module.exports = {
  csrfMissing,
  isDev,
  isTest,
};
