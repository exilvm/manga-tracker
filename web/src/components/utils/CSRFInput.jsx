import { Field } from 'react-final-form';
import { useCSRF } from '../../utils/csrf';

export default function CSRFInput() {
  const csrf = useCSRF();
  return (
    <Field
      name='_csrf'
      component='input'
      type='hidden'
      defaultValue={csrf}
    />
  );
}
