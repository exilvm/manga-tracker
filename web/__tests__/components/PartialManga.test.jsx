import React from 'react';
import renderer from 'react-test-renderer';
import PartialManga from '../../src/components/PartialManga';
import { emptyFullManga as emptyManga, fullManga as manga } from '../constants';

import { mockUTCDates } from '../utils';


describe('Partial manga should render correctly', () => {
  mockUTCDates();

  it('Should render correctly with no input', () => {
    const tree = renderer
      .create(<PartialManga />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Should render correctly with input', () => {
    const tree = renderer
      .create(<PartialManga mangaData={manga.manga} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Should render correctly with minimal input', () => {
    const tree = renderer
      .create(<PartialManga mangaData={emptyManga.manga} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Should render correctly when showId is true', () => {
    const tree = renderer
      .create(<PartialManga mangaData={manga.manga} showId />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
