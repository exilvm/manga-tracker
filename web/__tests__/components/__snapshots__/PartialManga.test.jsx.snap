// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Partial manga should render correctly Should render correctly when showId is true 1`] = `
<div>
  <h5
    className="MuiTypography-root makeStyles-title-1 MuiTypography-h5"
  >
    Dr. STONE
  </h5>
  <div
    className="makeStyles-details-3"
  >
    <a
      href="https://myanimelist.net/manga/103897"
      rel="noopener noreferrer"
      target="_blank"
    >
      <img
        alt="Dr. STONE"
        className="makeStyles-thumbnail-2"
        src="https://mangadex.org/images/manga/20882.jpg?1585634146"
      />
    </a>
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-justify-xs-space-between"
    >
      <table
        className="makeStyles-infoTable-5"
      >
        <tbody>
          <tr>
            <td>
              <p
                className="MuiTypography-root MuiTypography-body1"
              >
                Manga id:
              </p>
            </td>
            <td>
              <p
                className="MuiTypography-root makeStyles-detailText-4 MuiTypography-body1"
              >
                1
              </p>
            </td>
          </tr>
          <tr>
            <td>
              <p
                className="MuiTypography-root MuiTypography-body1"
              >
                Latest release:
              </p>
            </td>
            <td>
              <p
                aria-describedby={null}
                className="MuiTypography-root makeStyles-detailText-4 MuiTypography-body1"
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                title="Sun, 05 Jul 2020 16:00:00 GMT"
              >
                Jul 5th 2020, 16:00 - 1 day ago
              </p>
            </td>
          </tr>
          <tr>
            <td>
              <p
                className="MuiTypography-root MuiTypography-body1"
              >
                Estimated release interval:
              </p>
            </td>
            <td>
              <p
                className="MuiTypography-root makeStyles-detailText-4 MuiTypography-body1"
              >
                7 days 0 hours
              </p>
            </td>
          </tr>
          <tr>
            <td>
              <p
                className="MuiTypography-root MuiTypography-body1"
              >
                Estimated next release:
              </p>
            </td>
            <td>
              <p
                className="MuiTypography-root makeStyles-detailText-4 MuiTypography-body1"
              >
                Jul 12th 2020, 16:00
              </p>
            </td>
          </tr>
          <tr>
            <td>
              <p
                className="MuiTypography-root MuiTypography-body1"
              >
                Latest chapter:
              </p>
            </td>
            <td>
              <p
                className="MuiTypography-root makeStyles-detailText-4 MuiTypography-body1"
              >
                157
              </p>
            </td>
          </tr>
        </tbody>
      </table>
      <div
        className="makeStyles-sourceList-6"
      >
        <nav
          aria-labelledby="nested-list-subheader"
          className="MuiList-root makeStyles-root-8  MuiList-padding"
        >
          <div
            aria-disabled={false}
            className="MuiButtonBase-root MuiListItem-root makeStyles-mainItem-9 MuiListItem-gutters MuiListItem-button"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            <div
              className="MuiListItemText-root makeStyles-listOpener-11"
            >
              <span
                className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Manga sources
              </span>
            </div>
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
            <span
              className="MuiTouchRipple-root"
            />
          </div>
        </nav>
      </div>
    </div>
  </div>
</div>
`;

exports[`Partial manga should render correctly Should render correctly with input 1`] = `
<div>
  <h5
    className="MuiTypography-root makeStyles-title-1 MuiTypography-h5"
  >
    Dr. STONE
  </h5>
  <div
    className="makeStyles-details-3"
  >
    <a
      href="https://myanimelist.net/manga/103897"
      rel="noopener noreferrer"
      target="_blank"
    >
      <img
        alt="Dr. STONE"
        className="makeStyles-thumbnail-2"
        src="https://mangadex.org/images/manga/20882.jpg?1585634146"
      />
    </a>
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-justify-xs-space-between"
    >
      <table
        className="makeStyles-infoTable-5"
      >
        <tbody>
          <tr>
            <td>
              <p
                className="MuiTypography-root MuiTypography-body1"
              >
                Latest release:
              </p>
            </td>
            <td>
              <p
                aria-describedby={null}
                className="MuiTypography-root makeStyles-detailText-4 MuiTypography-body1"
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                title="Sun, 05 Jul 2020 16:00:00 GMT"
              >
                Jul 5th 2020, 16:00 - 1 day ago
              </p>
            </td>
          </tr>
          <tr>
            <td>
              <p
                className="MuiTypography-root MuiTypography-body1"
              >
                Estimated release interval:
              </p>
            </td>
            <td>
              <p
                className="MuiTypography-root makeStyles-detailText-4 MuiTypography-body1"
              >
                7 days 0 hours
              </p>
            </td>
          </tr>
          <tr>
            <td>
              <p
                className="MuiTypography-root MuiTypography-body1"
              >
                Estimated next release:
              </p>
            </td>
            <td>
              <p
                className="MuiTypography-root makeStyles-detailText-4 MuiTypography-body1"
              >
                Jul 12th 2020, 16:00
              </p>
            </td>
          </tr>
          <tr>
            <td>
              <p
                className="MuiTypography-root MuiTypography-body1"
              >
                Latest chapter:
              </p>
            </td>
            <td>
              <p
                className="MuiTypography-root makeStyles-detailText-4 MuiTypography-body1"
              >
                157
              </p>
            </td>
          </tr>
        </tbody>
      </table>
      <div
        className="makeStyles-sourceList-6"
      >
        <nav
          aria-labelledby="nested-list-subheader"
          className="MuiList-root makeStyles-root-8  MuiList-padding"
        >
          <div
            aria-disabled={false}
            className="MuiButtonBase-root MuiListItem-root makeStyles-mainItem-9 MuiListItem-gutters MuiListItem-button"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            <div
              className="MuiListItemText-root makeStyles-listOpener-11"
            >
              <span
                className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Manga sources
              </span>
            </div>
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
            <span
              className="MuiTouchRipple-root"
            />
          </div>
        </nav>
      </div>
    </div>
  </div>
</div>
`;

exports[`Partial manga should render correctly Should render correctly with minimal input 1`] = `
<div>
  <h5
    className="MuiTypography-root makeStyles-title-1 MuiTypography-h5"
  >
    Dr. STONE
  </h5>
  <div
    className="makeStyles-details-3"
  >
    <a
      href={null}
      rel="noopener noreferrer"
      target="_blank"
    >
      <img
        alt="Dr. STONE"
        className="makeStyles-thumbnail-2"
        src={null}
      />
    </a>
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-justify-xs-space-between"
    >
      <table
        className="makeStyles-infoTable-5"
      >
        <tbody>
          <tr>
            <td>
              <p
                className="MuiTypography-root MuiTypography-body1"
              >
                Latest release:
              </p>
            </td>
            <td>
              <p
                aria-describedby={null}
                className="MuiTypography-root makeStyles-detailText-4 MuiTypography-body1"
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                title="Unknown"
              >
                Unknown
              </p>
            </td>
          </tr>
          <tr>
            <td>
              <p
                className="MuiTypography-root MuiTypography-body1"
              >
                Estimated release interval:
              </p>
            </td>
            <td>
              <p
                className="MuiTypography-root makeStyles-detailText-4 MuiTypography-body1"
              >
                Unknown
              </p>
            </td>
          </tr>
          <tr>
            <td>
              <p
                className="MuiTypography-root MuiTypography-body1"
              >
                Estimated next release:
              </p>
            </td>
            <td>
              <p
                className="MuiTypography-root makeStyles-detailText-4 MuiTypography-body1"
              >
                Unknown
              </p>
            </td>
          </tr>
          <tr>
            <td>
              <p
                className="MuiTypography-root MuiTypography-body1"
              >
                Latest chapter:
              </p>
            </td>
            <td>
              <p
                className="MuiTypography-root makeStyles-detailText-4 MuiTypography-body1"
              >
                Unknown
              </p>
            </td>
          </tr>
        </tbody>
      </table>
      <div
        className="makeStyles-sourceList-6"
      >
        <nav
          aria-labelledby="nested-list-subheader"
          className="MuiList-root makeStyles-root-8  MuiList-padding"
        >
          <div
            aria-disabled={false}
            className="MuiButtonBase-root MuiListItem-root makeStyles-mainItem-9 MuiListItem-gutters MuiListItem-button"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            <div
              className="MuiListItemText-root makeStyles-listOpener-11"
            >
              <span
                className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Manga sources
              </span>
            </div>
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
            <span
              className="MuiTouchRipple-root"
            />
          </div>
        </nav>
      </div>
    </div>
  </div>
</div>
`;

exports[`Partial manga should render correctly Should render correctly with no input 1`] = `null`;
